private/buf/bufsync/syncer.go

```diff
@@ -94,508 +76,687 @@ func newSyncer(
-	modIdentity := moduleIdentity.IdentityString()
-	// check local cache first
-	if syncedModuleCommits, ok := s.modulesIdentitiesToCommitsSyncedCache[modIdentity]; ok {
-		if _, commitSynced := syncedModuleCommits[commitHash.Hex()]; commitSynced {
+		manifestDigest := manifestBlob.Digest().String()
+		if manifestDigest == bsrCommitToMatch.ManifestDigest {
+			// Content-match found; select this commit.
 			return true, nil
 		}
-	}
-	// not in the cache, request BSR check
-	commitSynced, err := s.handler.IsGitCommitSynced(ctx, moduleIdentity, commitHash)
+		return false, nil
+	})
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	if commitSynced {
-		// populate local cache
-		if s.modulesIdentitiesToCommitsSyncedCache[modIdentity] == nil {
-			s.modulesIdentitiesToCommitsSyncedCache[modIdentity] = make(map[string]struct{})
-		}
-		s.modulesIdentitiesToCommitsSyncedCache[modIdentity][commitHash.Hex()] = struct{}{}
+	if matched != nil {
+		s.logger.Debug(
+			"content matched to commit",
+			zap.String("moduleDir", moduleDir),
+			zap.String("moduleIdentity", moduleIdentity.IdentityString()),
+			zap.String("branch", branch),
+			zap.String("gitHash", matched.Hex()),
+			zap.String("bsrCommitID", bsrCommitToMatch.Id),
+		)
+		return walked, nil
 	}
-	return commitSynced, nil
+	// Select only the HEAD of the branch.
+	return []git.Hash{head}, nil
 }
 
-// readModule returns a module that has a name and builds correctly given a commit and a module
-// directory.
-func (s *syncer) readModuleAt(
+// protectSyncedModuleBranch protects a synced branch according to the following rules:
+//   - If a protected branch is _behind_ the BSR, this is allowed.
+//   - If a protected branch is _ahead of_ the BSR, this is also allowed.
+//   - If a protected branch is both _ahead of_ and _behind_ the BSR, this is disallowed.
+//
+// This must only be called for protected branches.
+func (s *syncer) protectSyncedModuleBranch(
 	ctx context.Context,
-	commit git.Commit,
-	moduleDir string,
-) (*bufmodulebuild.BuiltModule, error) {
-	commitBucket, err := s.storageGitProvider.NewReadBucket(commit.Tree(), storagegit.ReadBucketWithSymlinksIfSupported())
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branch string,
+) error {
+	syncPoint, err := s.resolveSyncPointForProtectedBranch(ctx, moduleIdentity, branch)
 	if err != nil {
-		return nil, fmt.Errorf("new read bucket: %w", err)
+		return err
 	}
-	moduleBucket := storage.MapReadBucket(commitBucket, storage.MapOnPrefix(moduleDir))
-	foundModule, err := bufconfig.ExistingConfigFilePath(ctx, moduleBucket)
-	if err != nil {
-		return nil, fmt.Errorf("looking for an existing config file path: %w", err)
+	if syncPoint != nil {
+		// Branch has never been synced, there is nothing to protected against.
+		return nil
 	}
-	if foundModule == "" {
-		// No module at this commit.
-		return nil, nil
+	if containsSyncPoint, err := s.branchContainsCommit(branch, syncPoint); err != nil {
```

Comment:
I might be wrong, but if you're doing this here, you don't need to check the sync point you're receiving from the BSR exists in the Git repo. This will return false anyway.

If that's the case, then the whole `resolveSyncPointForProtectedBranch` can be gone, and you could use the s.handler.SyncPoint` here.

Reply:
Also, if the sync point is ahead of us, it won't exist in the local Git repo. I think definitely we can kill that func with that validation there, and just use the handler here.

Reply:
You're right, good catch, I killed it.